### ISSUE TITLE
Proposal: retry try statements

### ISSUE THREAD

**Issue by GeirGrusom** [dotnet/csharplang#127]
Sometimes an exception has been thrown because of a temporary state, such as network error. In order to replay the try block in C# 6 you would implement a loop. This will cause the finally block to run for every retry which may be undesirable. You cannot jump from a catch to a try scope since the label is out of scope.

I propose that a standalone `try;` replays the try block. Alternatively, a new keyword `retry` could be used.

```csharp
var resource = GetResource()
int retryCount = 0;
try
{
    targetInfo = await resource.GetNetworkUserInfo();
}
catch(TimeoutException) when (++retryCount <= 3)
{
    await Task.Delay(10000);
    try;
}
finally
{
    resource.Dispose();
}
```

The finally block is only executed once regardless of how many times the block was retried.

**Comment by Logerfo** [comment]
```
int retryCount = 0;
try
{
    targetInfo = await GetNetworkUserInfo();
}
catch(TimeoutException) when (++retryCount <= 3)
{
    await Task.Delay(10000);
    try
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch(Exception)
    {
        try; //which "try" block does this point to?
    }
}
```

**Comment by orthoxerox** [comment]
@Logerfo What loop does `break;` point to?

**Comment by Logerfo** [comment]
@orthoxerox unfortunately, c# break doesn't support argument like [PHP](http://php.net/manual/pt_BR/control-structures.break.php) does. We could have try labels, though:

```
int retryCount = 0;
try first
{
    targetInfo = await GetNetworkUserInfo();
}
catch(TimeoutException) when (++retryCount <= 3)
{
    await Task.Delay(10000);
    try second
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch(Exception)
    {
        try first;
    }
}
```

**Comment by GeraudFabien** [comment]
I don't understand why put the "when (++retryCount <= 3)" on the "catch" line. If you have many catch each catch must have the same condition and if not it will be hard to understand the flow and debug (I thinks).
```
int retryCount = 0;
try when (++retryCount <= 3) // while
{
    if(retryCount == 0){
        targetInfo = await GetNetworkUserInfo();
    }
    
    targetInfo = await GetNetworkUserInfo();
}
catch(TimeoutException) 
{
    await Task.Delay(10000);
}
catch(OtherException)
{
    // or "break try;" or "try;"
    break; 
}
```

**Comment by GeirGrusom** [comment]
@GeraudFabien 

> I don't understand why put the "when (++retryCount <= 3)" on the "catch" line.

Because that is what by C# supports at this point, and different exceptions may have different retry conditions.

Take this hypothetical, wrong and probably far-fetched example:

```csharp
string connId = "db1";
int count = 0;

try
{
  db.ConnectionString = GetConnectionString(connId);
  await db.OpenAsync();
}
catch(TimeoutException) when (++count <= 3)
{
  retry;
}
catch(BadCredentialsException) when (connId != "db2")
{
  connId = "db2";
  retry;
}
```

@Logerfo 

> We could have try labels, though

We could, but I don't really think it's necessary. It should probably go in a different proposal as well.

**Comment by alrz** [comment]
You are coupling the actual procedure and retry logic. you probably should be using a separate method/type to handle that logic. However I can see this could be covered by modifying generators (#122)


```cs
[[Retry(3]]
{
  DoSomething();
}
```
or something like that.

**Comment by GeirGrusom** [comment]
> You are coupling the actual procedure and retry logic. you probably should be using a separate method/type to handle that logic.
 
You are making a sweeping comment on code structure that I don't think is really relevant here.

> However I can see this could be covered by modifying generators

```csharp
[[Retry(3)]]
{
  DoSomething();
}
```

You try to fit my simple example without giving any thought to the point that some exceptions are recoverable. I used `TimeoutException` as an example where it just waited for whatever to resolve itself. But it may be updating a database record, or doing anything else before retrying.

Also, this feature is fairly simple. It is no more complicated to implement than `continue`. Code generators are however are not simple to implement, and your example would introduce a lot of cognitive load.

**Comment by scottdorman** [comment]
I think I'd rather see this as a `retry` keyword. 

In @Logerfo's example (shown below for reference)

```c#
int retryCount = 0;
try
{
    targetInfo = await GetNetworkUserInfo();
}
catch(TimeoutException) when (++retryCount <= 3)
{
    await Task.Delay(10000);
    try
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch(Exception)
    {
        try; //which "try" block does this point to?
    }
}
```

The `try; //which "try" block does this point to?`, it should retry the associated `try` block that it's attached to. In this case, it would retry in the inner most `try` block.

If we went the path of using exception filters, I'd like to see `retryCount` be an implicit variable (in the same way that `value` is for properties).

**Comment by YaakovDavis** [comment]
I propose `retry` instead of `try`; should be easier to read and to analyze.

```
try
{
   ...
}
catch
{
   if(condition())
      retry;
}

**Comment by Logerfo** [comment]
@scottdorman both `retry` and `retryCount` would need to be new reserved keywords, which is bad for retrocompatibility. Also, I don't think `try(n)` is reliable, considering the `try` statement can appear many times in the same `catch` block, like that:
```C#
int retryCount = 0;
try 
{        
    targetInfo = await GetNetworkUserInfo();
}
catch(TimeoutException)
{
    await Task.Delay(10000);
    if(retryCount++ > 0)
        try(2); //will be hit at the first try
    else
        try(3); //will be hit at the second try
}
```

**Comment by scottdorman** [comment]
@Logerfo 

> both retry and retryCount would need to be new reserved keywords

`retry` would need to be a new contextual keyword. I don't think `retryCount` would since it would be compiler generated/emitted it can be named whatever we want, in the same way autoproperties generate an implicit backing field.

You also wouldn't have code like 

```
if(retryCount++ > 0)
        try(2); //will be hit at the first try
    else
        try(3); //will be hit at the second try
```

`retryCount` isn't an accessible variable (again, just like the generated backing field for an autoprop) so you wouldn't be able to write code like that. The `try(n)` is simply providing the number of times to retry. In my example, it would get expanded to something like this:

```c#
{
   var ___retryCountBackingField = 0;
   try
   {
      targetInfo = await GetNetworkUserInfo();
   }
   catch(TimeoutException) 
   {
      await Task.Delay(10000);
      try
      {
         targetInfo = await GetNetworkUserInfo();
      }
      catch(Exception) when (++___retryCountBackingField <= 3)
      {
         retry(3);
      }
   }
}
```

**Comment by Logerfo** [comment]
> retryCount isn't an accessible variable

yes it is, I've declared it:
```C#
int retryCount = 0;
```

I've used this variable just to exemplify a code flow possibility involving multiple `try(n)` with different `n` being called on each retry.

**Comment by scottdorman** [comment]
@Logerfo In your example, yes, you've declared it. In my suggestion, I'm saying that you wouldn't have it/need it as it would be automatically generated. Since it's not an accessible variable, you wouldn't be able to have code like you're suggesting. Yes, you could still have multiple `retry(n)` statements if you had multiple catch handlers, but each of those `retry` statements would generate a different backing field (which is what It think you'd want anyway in your example since different conditions could have different retry strategies) and each of the associated `catch` statements would have the appropriate exception filter applied to it.

**Comment by Logerfo** [comment]
@scottdorman what I'm trying to say is that the condition to flow to different `try(n)` statements does not necessarily depends on the retry count. Also, I'm talking about multiple `try(n)` statements in the same catch block. What would happen when the code reach `try(2)` on the first attempt and `try(3)` on the second attempt? That's why I think the parameter doesn't fit.

**Comment by scottdorman** [comment]
Here is another example:

```c#
{
    var ___retryCountBackingField1 = 0;
    var ___retryCountBackingField2 = 0;
    var ___retryCountBackingField3 = 0;

    try // outer
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch (TimeoutException) when (++___retryCountBackingField1 <= 4)
    {
        await Task.Delay(10000);
        try // inner
        {
            targetInfo = await GetNetworkUserInfo();
        }
        catch (NullReferenceException) when (++___retryCountBackingField2 <= 3)
        {
            retry(3); // retries the "inner" try block up to 3 times.
        }
        catch (System.Net.HttpListenerException) when (++___retryCountBackingField3 <= 2)
        {
            retry(2); // retries the "inner" try block up to 2 times.
        }

        retry(4); // retries the "outer" try block up to 4 times.
    }
}
```

This still has two `try` blocks, one nested inside the other, and the inner most `try` block has multiple `catch` handlers. I think this covers the scenario you described and I think it will still work properly. If the "inner" `try` block catches a `NullReferenceException` it will retry. If, on that retry, it catches another `NullReferenceException`, it will retry again, and it will keep retrying as long as it continues getting `NullReferenceException` up to 3 times. If it gets an `HttpListenerException` during one of those tries, it will still retry that, but up to 2 times (as long as it keeps getting an `HttpListenerException`).

So, the retry and counter variable is specific to the exception type your handling and only gets incremented/retried when you encounter that exception. In this example, if the inner most `try` block threw something other than `NullReferenceException` or `HttpListenerException` neither of those `catch` handlers would be hit, so the `retry(4)` would be the only one that comes in to play.

**Comment by Logerfo** [comment]
@scottdorman You're example seems to work, but it doesn't cover the scenario I described:
> I'm talking about multiple try(n) statements in the **same** catch block.

**Comment by scottdorman** [comment]
@Logerfo Why would you need multiple retries from within the same catch handler?

**Comment by chrisaut** [comment]
Give theres a fairly simple alternative, which is to just use goto, I don't think this warrants the work required to make this a reality.

```c#
            int retryCount = 0;
            retry: try
            {
                targetInfo = await GetNetworkUserInfo();
            }
            catch (TimeoutException) when (++retryCount <= 3)
            {
                await Task.Delay(10000);
                goto retry;
            }
```

**Comment by Logerfo** [comment]
@scottdorman code flow. It may or may not try again, given a certain logic, which may need more then one statement.

@chrisaut in your code, if present, a `finally` block would be executed? Whatever the answer is, would we want it to execute if a "retry" (or whatever) statement is developed?

**Comment by tannergooding** [comment]
@chrisaut, but we can't use goto, because raptors ;) https://xkcd.com/292/

**Comment by chrisaut** [comment]
@Logerfo yes, finally would be executed because it jumps outside of the try/catch/finally block.

Are you expecting your ```retry``` statement to skip executing finally? What if inside the try you take on some exclusive resource, surely you would want the finally to be executed so that you can retry from a clean state, no?

**Comment by scottdorman** [comment]
@Logerfo So you're looking for something similar to this?

```c#
{
    var ___retryCountBackingField1 = 0;
    var ___retryCountBackingField2 = 0;

    try // outer
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch (SqlException e) when (++___retryCountBackingField1 <= 4 || ++___retryCountBackingField2 <= 2)
    {
        if (e.ErrorCode < 200)
        {
            retry(2);
        }

        retry(4); // retries the "outer" try block up to 4 times.
    }
}
```

**Comment by vladd** [comment]
What is a real benefit of this form over simple

```
for (int iteration = 0; iteration < 3; iteration++)
{
    try
    {
        targetInfo = await GetNetworkUserInfo();
        break;
    }
    catch(TimeoutException) when (iteration < 2)
    {
        await Task.Delay(10000);
    }
}
```

?

With the simple form, I can make iteration condition more complicated (and don't need to squeeze it into `when` clause), I can replace `for` with `while`, I can make it iterate through a collection, and this form is known to everyone and not actually that long.

**Comment by GeirGrusom** [comment]
You would need an error flag or a break in the finally block or you would unconditionally rerun the try block.

So I would say that it makes it simpler to express these kinds of things correctly.

**Comment by Logerfo** [comment]
@chrisaut I see. I didn't say I was expecting it to skip, I was really just asking.

@scottdorman Exactly! What if the first 3 attempts returns an error code above 200 and the 4th returns an error code below 200? I can't see the expected behavior.

**Comment by scottdorman** [comment]
@Logerfo 

> What if the first 3 attempts returns an error code above 200 and the 4th returns an error code below 200? 

If the 4th attempt returns an error code below 200, then it will hit the `if` block, and retry again, up to two more times. In this scenario, I think you could theoretically retry up to 6 times.

**Comment by Logerfo** [comment]
@scottdorman IMO that's way too complex, considering the parameter only accepts const values. Otherwise, it's way too complex a thousand times.

**Comment by scottdorman** [comment]
@Logerfo Complex how? Which parameter? The example code should fit your scenario and do what it's supposed to do, correct?

If I were writing this, I'd expect to write code like this:

```c#
try 
{
    targetInfo = await GetNetworkUserInfo();
}
catch (SqlException e) when (e.ErrorCode < 200)
{
    retry(2);
}
catch (SqlException e) when (e.ErrorCode >= 200)
{ 
    retry(4); 
}
```

and have the compiler expand that to something like this:

```c#
{
    var ___retryCountBackingField1 = 0;
    var ___retryCountBackingField2 = 0;

    try
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch (SqlException e) when (e.ErrorCode < 200 && ++___retryCountBackingField1 <= 2)
    {
        retry(2);
    }
    catch (SqlException e) when (e.ErrorCode >= 200 && ++___retryCountBackingField2 <= 4)
    { 
        retry(4);
    }
}
```

**Comment by jnm2** [comment]
I like `retry` better than `try`, but I'd also have liked `rethrow` and we have `throw`.

**Comment by Logerfo** [comment]
@scottdorman please keep in mind that not every logic fit the `when` statement. About the parameter, check out this:

```c#
int rnd() => new Random().Next(1, 10);
try 
{
    targetInfo = await GetNetworkUserInfo();
}
catch (SqlException e) when (e.ErrorCode < 200)
{
    retry(rnd());
}
catch (SqlException e) when (e.ErrorCode >= 200)
{ 
    retry(rnd()); 
}
```

**Comment by scottdorman** [comment]
@Logerfo You'd want an arbitrarily random number of retries? When would the number of retries ever be anything other than a constant value? Exception filters can contain any synchronous expression that returns a `bool`, so your logic inside the `when` can be whatever you want/need it to be.

**Comment by Logerfo** [comment]
@scottdorman 
> You'd want an arbitrarily random number of retries?

Random was just an example for a variable number.
> When would the number of retries ever be anything other than a constant value?

Maybe the user could set it?
> Exception filters can contain any synchronous expression that returns a bool, so your logic inside the when can be whatever you want/need it to be.

Sorry, you're right.

**Comment by scottdorman** [comment]
@Logerfo 

> Random was just an example for a variable number. Maybe the user could set it?

Fair enough. I could see a scenario where the number of retries is user defined, in which case accepting only a constant value for `retry` wouldn't work. However, could `retry` have two overloads (effectively), one which takes a constant value and another which takes an expression that returns an integer? That would support both scenarios.

**Comment by Logerfo** [comment]
@scottdorman only if the expression is evaluated only once

**Comment by pharring** [comment]
For the sake of exposition, please would you show the compiler-generated code using ```goto``` instead of a hypothetical ```retry``` primitive? (There isn't really a retry IL instruction or CompilerServices primitive is there?)

**Comment by eyalsk** [comment]
@Logerfo In many cases you would want to allow the client to pass the amount of retries it needs to perform so what's the benefit of this over a goto or recursive approach?

```
async Task DoAsync(int retryCount = 0)
{
	try
	{
		await GetNetworkUserInfo();
	}
	catch (TimeoutException e) when (++retryCount <= 3)
	{
		await Task.Delay(2000);
		
		await DoAsync(retryCount);
	}
}
```

**Comment by Logerfo** [comment]
@eyalsk AFAIK your code is tail recursive, so will only be optimized using 64bit JIT. A retry statement could do this with jump instructions, way faster.

**Comment by scottdorman** [comment]
@pharring I think this is what you're looking for, yes?

Given this code:

```c#
try
{
    targetInfo = await GetNetworkUserInfo();
}
catch (SqlException e) when (e.ErrorCode < 200)
{
    retry(2);
}
catch (SqlException e) when (e.ErrorCode >= 200)
{
    retry(4); // retries the "outer" try block up to 4 times.
}
```

The compiler expanded code could (roughly) be

```c#
{
    var ___retryCountBackingField1 = 0;
    var ___retryCountBackingField2 = 0;

    ___try1:  try
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch (SqlException e) when (e.ErrorCode < 200 && ++___retryCountBackingField1 <= 2)
    {
        goto ___try1;
    }
    catch (SqlException e) when (e.ErrorCode >= 200 && ++___retryCountBackingField2 <= 4)
    {
        goto ___try1;
    }
}
```

**Comment by scottdorman** [comment]
@Logerfo Why would you want to restrict the expression to only being evaluated once?

**Comment by pharring** [comment]
Nit: Consider initializing the backing field to the retry count (instead of zero) then decrement it and compare to zero. That way the constant is not part of the test.

**Comment by YaakovDavis** [comment]
In a given app, retry logic could be based on many other factors other than *count*.  
Even if `retry(n)` is doable logically, it serves such a narrow use-case, that isn't syntax-worthy IMO.

**Comment by scottdorman** [comment]
@pharring Nice. So, I think that would change it to look like this:

Given this code:

```c#
try
{
    targetInfo = await GetNetworkUserInfo();
}
catch (SqlException e) when (e.ErrorCode < 200)
{
    retry(2);
}
catch (SqlException e) when (e.ErrorCode >= 200)
{
    retry(4); // retries the "outer" try block up to 4 times.
}
```

The compiler expanded code could (roughly) be

```c#
{
    var ___retryCountBackingField1 = 2;
    var ___retryCountBackingField2 = 4;

    ___try1:  try
    {
        targetInfo = await GetNetworkUserInfo();
    }
    catch (SqlException e) when (e.ErrorCode < 200 && --___retryCountBackingField1 > 0)
    {
        goto ___try1;
    }
    catch (SqlException e) when (e.ErrorCode >= 200 && --___retryCountBackingField2 > 0)
    {
        goto ___try1;
    }
}
```

**Comment by eyalsk** [comment]
@YaakovDavis Good point although the original proposal which I still disagree with, does not take any arguments.

**Comment by YaakovDavis** [comment]
@eyalsk I was referring the `n` arg specifically. 
I'd say `retry;` without args is somewhat useful, though not necessarily worth its cost.

**Comment by Thaina** [comment]
Instead of adding `retry`. Shouldn't we just resurrect `goto`

`goto try;` maybe?

**Comment by GeirGrusom** [comment]
Does anyone know if it's possible in IL to make a try..catch..retry that only calls the finally block once it actually leaves the entire `try..catch`? Rather than calling `finally` every time it gives control back to the `try` block.

**Comment by Thaina** [comment]
Actually we could implement this by convert the whole block into while block

I mean in any keyword we should to retry we would write it like this
```C#
try
{
     DoSomething();
}
catch(Exception e)
{
    if(condition)
       goto try;
}
finally
{
}
```

It would be transpiled into

```C#
try
{
     Exception e;
     do
     {
        try
        {
             DoSomething();
        }
        catch(Exception ex){ e = ex; }
     }
     while(condition);
}
finally
{
}
```

**Comment by Logerfo** [comment]
Considering the `finally` block is always executed once the code leaves the try/catch block, as this is the default behavior with currently language specs, the compiler already translate the `goto` to a `while` block, [as you can see](http://tryroslyn.azurewebsites.net/#b:master/f:r/K4Zwlgdg5gBAygTxAFwKYFsDcAoADsAIwBswBjGUogQxBBgGEYBvbGNmfYsmANwHswAExgBZABQBKZq3azIyGFRgBeGACYcs2dQKoiALhjIATghlbpFi1QDUNzVYC+5raSrJSACxeyWVrVB8yHwwOnoOFs7+AGaQVEREZv5+/uxUALTpEVpR7M6OQA==).

**Comment by GeirGrusom** [comment]
@Logerfo that's just the decompilers interpretation of the code. A `do { } while(condition);` is indistinguishable from a `label: if(condition) goto label;` after compilation.

**Comment by jnm2** [comment]
I see why it's desirable to execute the finally block only once, because typically it looks like this"

```
BeginX();
try
{
    DoSomethingThatMayOrMayNotBeExpectedToFail();
}
finally
{
    EndX();
}
```

No matter what `BeginX` and `EndX` do, you want them to be executed in a balanced way: exactly one `BeginX` for every `EndX`.

**Comment by HaloFour** [comment]
Seems like a lot of voodoo for relatively uncommon scenarios.

It also runs the risk of baking in too much concept around how exception handling should work.  I've found it exceptionally rare that I care how many *times* that something has been tried, but I might care how long I've been trying depending on the circumstances.  Here the compiler would have to make the decision as to how retries are applied on different `catch` blocks, which prevents the developer from doing so.

**Comment by GeirGrusom** [comment]
I've been a little bit disappointed by how much we've been focusing on my example rather than the proposal itself. The idea here isn't specifically to solve the 'retry count' problem, but more generally be able to re-run a try expression for whatever reason.

In my opinion the finally block should only be run *once* when the code runs out of the try..catch, not for every retry. You can't express this in C# currently since placing a label in the try block is inaccessible from the catch block.

**Comment by eyalsk** [comment]
@GeirGrusom How often you find yourself handle exceptions and how often you need to recover from them?

**Comment by GeirGrusom** [comment]
@eyalsk I don't actually do it that often but I don't use `continue` that often either, or `dynamic`. I also don't find pattern matching that useful in practice (Kotlin has it) besides matching on types and specific values. All examples are always really contrived to show the usefulness of the feature. Still there is a huge effort in implementing it in C# even though I at least think it's a feature that's more awesome on paper than in practice.

This feature on the other hand is very small, and it is at least able to do something you otherwise can't express without rearranging code. You can't loop without rerunning the `finally` and you can't `goto` inside the `try` block.

**Comment by eyalsk** [comment]
@GeirGrusom 

> Not every proposal has to be something you use every day. Features that saves you from writing awkward code once in a while also has merit.

Well, I don't disagree, I just think that if you don't use it everyday and use it once in a while then the feature should have a much bigger impact for them to consider it but I could be wrong. :)

**Comment by GeirGrusom** [comment]
Sure. I'm not exactly heavily invested in this. I implemented this feature in a DSL I made but in that language it's a more common case because it was designed to give a simple solution for when credit card charge fails on subscriptions. C# can't do that but I still thought it's a little odd to have `try` without a `retry`.

> Well kids, you tried your best and failed miserably. The lesson is: never try.

So I'll readily admit that it's not really a huge issue, and I wouldn't exactly shed a tear if it's not implemented. It just seems like it's a fairly obvious piece that's missing among `continue`,  `break` and `throw` so I thought I would suggest it.

**Comment by ashmind** [comment]
Is there something way we can move the original proposal forward?
Just a `retry` without any counts, goes back to `try`.

It's not that common, but just ran into it again and it's kind of awkward to implement otherwise.
I feel I might code retries more often if it didn't require loops (+extra level of nesting) or a helper.

**Comment by fuzzzeee** [comment]
I would like to see this become part of the language.. but in the meantime you can roll your own.

This class can be used to wrap up the retry logic and extended to create reusable retry configurations.

```
    public class TryHard
    {
        private int _attempts = 3;

        /// <summary>
        /// The total number of attempts to make. Defaults to 3
        /// </summary>
        public int Attempts
        {
            get => _attempts;
            set
            {
                if (Attempts != value)
                {
                    if (value < 1)
                        throw new ArgumentOutOfRangeException(nameof(value), value, "Attempts cannot be less than 1");
                    _attempts = value;
                }
            }
        }

        /// <summary>
        /// The conditions which will result in another attempt
        /// </summary>
        public virtual ICollection<Func<Exception, bool>> ExceptionFilters { get; } = new List<Func<Exception, bool>>();

        /// <summary>
        /// The action to take between attempts. Defaults to an exponential delay
        /// </summary>
        public virtual Action<int> BetweenAttempts { get; set; } = n => Thread.Sleep(TimeSpan.FromSeconds(n * n));

        public void Attempt(Action action)
        {
            Exception lastException = null;
            for (var attempt = 0; attempt < Attempts; attempt++)
            {
                try
                {
                    action();
                    break;
                }
                catch (Exception e) when (ExceptionFilters.Any(f => f(e)))
                {
                    lastException = e;
                    BetweenAttempts(attempt + 1);
                }
            }
            if (lastException != null)
                throw lastException;
        }
    }

```

**Comment by bondsbw** [comment]
If the goal is to loop, we could have `try`/`while` in the spirit of `do`/`while`:

```c#
try
{
    ...
}
while catch (Exception ex)
{
    ...
}
finally { ... }
```

**Comment by bondsbw** [comment]
Three tries (explicit loop condition):
```c#
int x = 0;
try
{
   ...
}
while catch (Exception ex)
{
   ...
}
while (x < 3);
```

**Comment by bondsbw** [comment]
A few details on `try`/`while`:

- `finally` is only invoked once (upon any exit) and may not precede `while catch`
- Multiple `while catch` clauses are permitted
- Only one `while` with a condition is permitted
- Due to ambiguity, `while` with condition must either include a semicolon, or precede `finally`
  - _Maybe we don't allow `while` with semicolon... missing the semicolon could yield a valid but incorrect program_

**Comment by BackDoorManUC** [comment]
use https://github.com/App-vNext/Polly

**Comment by FiniteReality** [comment]
In my experience, most cases where something fails sporadically, especially when accessing network resources, are locations where you'd want to implement some sort of backoff, to prevent causing an effective denial of service attack. I don't see any easy way to implement that with this suggestion, that wouldn't be just as easy to write without it. As such, I don't think this a suitable feature for C# - it implements too little to be useful, in my opinion.

I feel the suggestion posted by @chrisaut is likely the best compromise here for those who want this - gotos are not necessarily evil when used correctly, and IMO, this is a perfect use case for it.

**Comment by GeirGrusom** [comment]
@FiniteReality 

> gotos are not necessarily evil when used correctly, and IMO, this is a perfect use case for it.

You cannot `goto` a label inside the try block, and jumping before it will cause the finalizer to execute.

**Comment by FiniteReality** [comment]
The latter case I didn't actually know about, but it makes sense now that I think about it. Either way, any type of loop would likely work just as well.

**Comment by mganss** [comment]
FWIW Ruby has a [`retry`](http://ruby-doc.com/docs/ProgrammingRuby/html/tut_exceptions.html#UB) keyword. Also the only language I know of that has builtin syntax for this scenario. Are there others?
